File: internal/ocr/image.go
package ocr

import (
	"bytes"
	"fmt"
	"image"
	"image/color"
	_ "image/gif"
	_ "image/jpeg"
	_ "image/png"
	"math"
	"net/http"

	"github.com/disintegration/imaging"
	_ "golang.org/x/image/webp"
)

const (
	MaxImageBytes = 10 << 20
	// maxPreparedBytes is the cap after JPEG quality. Dimensions stay as
	// photographed unless the file is still over this size.
	maxPreparedBytes = 3 << 19 // 1.5 MiB
	jpegQuality      = 82
	shrinkAttempts   = 16
)

func PrepareJPEG(raw []byte) ([]byte, error) {
	if len(raw) == 0 {
		return nil, ErrNoImage
	}
	if len(raw) > MaxImageBytes {
		return nil, fmt.Errorf("image must be 10 MB or smaller")
	}
	if !sniffImage(raw) {
		return nil, fmt.Errorf("image must be jpeg, png, webp, or gif")
	}
	img, err := imaging.Decode(bytes.NewReader(raw), imaging.AutoOrientation(true))
	if err != nil {
		return nil, fmt.Errorf("could not read the image")
	}
	img = flattenWhite(img)
	return fitJPEG(img, maxPreparedBytes)
}

func flattenWhite(img image.Image) image.Image {
	flat := imaging.New(img.Bounds().Dx(), img.Bounds().Dy(), color.White)
	return imaging.OverlayCenter(flat, img, 1)
}

func encodeJPEG(img image.Image) ([]byte, error) {
	var buf bytes.Buffer
	if err := imaging.Encode(&buf, img, imaging.JPEG, imaging.JPEGQuality(jpegQuality)); err != nil {
		return nil, err
	}
	return buf.Bytes(), nil
}

func fitJPEG(img image.Image, maxBytes int) ([]byte, error) {
	out, err := encodeJPEG(img)
	if err != nil {
		return nil, err
	}
	if len(out) <= maxBytes {
		return out, nil
	}

	origW := img.Bounds().Dx()
	origH := img.Bounds().Dy()
	scale := 1.0
	for i := 0; i < shrinkAttempts && len(out) > maxBytes; i++ {
		ratio := math.Sqrt(float64(maxBytes) / float64(len(out)))
		if ratio > 0.95 {
			ratio = 0.95
		}
		scale *= ratio
		w := int(math.Round(float64(origW) * scale))
		h := int(math.Round(float64(origH) * scale))
		if w < 1 {
			w = 1
		}
		if h < 1 {
			h = 1
		}
		out, err = encodeJPEG(imaging.Resize(img, w, h, imaging.Lanczos))
		if err != nil {
			return nil, err
		}
	}
	if len(out) > maxBytes {
		return nil, fmt.Errorf("could not compress the image under 1.5 MB")
	}
	return out, nil
}

func sniffImage(b []byte) bool {
	if len(b) >= 12 && string(b[:4]) == "RIFF" && string(b[8:12]) == "WEBP" {
		return true
	}
	switch http.DetectContentType(b) {
	case "image/jpeg", "image/png", "image/gif":
		return true
	}
	return false
}

File: src/ocr/image.ts
import sharp from 'sharp';
import { MaxImageBytes, NoImageError } from './types.js';

export const maxPreparedBytes = 3 << 19; // 1.5 MiB
const jpegQuality = 82;
const shrinkAttempts = 16;

export async function prepareJPEG(raw: Buffer): Promise<Buffer> {
  if (raw.length === 0) throw new NoImageError();
  if (raw.length > MaxImageBytes) throw new Error('image must be 10 MB or smaller');
  if (!sniffImage(raw)) throw new Error('image must be jpeg, png, webp, or gif');
  let flattened: Buffer;
  try {
    flattened = await flattenWhite(raw);
  } catch {
    throw new Error('could not read the image');
  }
  return fitJPEG(flattened, maxPreparedBytes);
}

export async function flattenWhite(raw: Buffer): Promise<Buffer> {
  return sharp(raw, { failOn: 'error' })
    .rotate()
    .flatten({ background: { r: 255, g: 255, b: 255 } })
    .png()
    .toBuffer();
}

export async function encodeJPEG(img: Buffer): Promise<Buffer> {
  return sharp(img).jpeg({ quality: jpegQuality }).toBuffer();
}

export async function fitJPEG(img: Buffer, maxBytes: number): Promise<Buffer> {
  let out = await encodeJPEG(img);
  if (out.length <= maxBytes) return out;

  const meta = await sharp(img).metadata();
  const origW = meta.width ?? 1;
  const origH = meta.height ?? 1;
  let scale = 1;
  for (let i = 0; i < shrinkAttempts && out.length > maxBytes; i++) {
    let ratio = Math.sqrt(maxBytes / out.length);
    if (ratio > 0.95) ratio = 0.95;
    scale *= ratio;
    let w = Math.round(origW * scale);
    let h = Math.round(origH * scale);
    if (w < 1) w = 1;
    if (h < 1) h = 1;
    out = await encodeJPEG(await sharp(img).resize(w, h, { kernel: 'lanczos3' }).toBuffer());
  }
  if (out.length > maxBytes) throw new Error('could not compress the image under 1.5 MB');
  return out;
}

export function sniffImage(b: Buffer): boolean {
  if (b.length >= 12 && b.subarray(0, 4).toString() === 'RIFF' && b.subarray(8, 12).toString() === 'WEBP') {
    return true;
  }
  if (b.length >= 3 && b[0] === 0xff && b[1] === 0xd8) return true;
  if (b.length >= 8 && b.subarray(0, 8).equals(Buffer.from([0x89, 0x50, 0x4e, 0x47, 0x0d, 0x0a, 0x1a, 0x0a]))) {
    return true;
  }
  if (b.length >= 6) {
    const head = b.subarray(0, 6).toString('ascii');
    if (head === 'GIF87a' || head === 'GIF89a') return true;
  }
  return false;
}
